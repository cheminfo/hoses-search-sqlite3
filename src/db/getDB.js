import fs, { existsSync, mkdirSync } from 'fs';
import { join } from 'path';

import sqLite from 'better-sqlite3';

let db;

export default function getDB() {
  const path = new URL('../../sqlite/', import.meta.url).pathname;
  console.log({ path })
  if (!db) {
    if (!existsSync(path)) {
      mkdirSync(path);
    }
    db = sqLite(join(path, 'db.sqlite'));
    // https://www.sqlite.org/wal.html
    // Activating WAL mode allows to get a speed improvement of 100x !!!
    db.pragma('journal_mode = WAL');
  }

  setInterval(() => {
    fs.stat(join(path, 'db.sqlite-wal'), (err, stat) => {
      if (err) {
        if (err.code !== 'ENOENT') throw err;
      } else if (stat.size > 100000000) {
        db.pragma('wal_checkpoint(RESTART)');
      }
    });
  }, 300000).unref();

  const createTable = `
CREATE TABLE IF NOT EXISTS molecules (
  id data_type PRIMARY KEY,
  idCode data_type TEXT,
  ssIndex0 data_type INTEGER,
  ssIndex1 data_type INTEGER,
  ssIndex2 data_type INTEGER,
  ssIndex3 data_type INTEGER,
  ssIndex4 data_type INTEGER,
  ssIndex5 data_type INTEGER,
  ssIndex6 data_type INTEGER,
  ssIndex7 data_type INTEGER
);
CREATE INDEX IF NOT EXISTS ssIndex0 ON molecules(ssIndex0);
CREATE INDEX IF NOT EXISTS ssIndex1 ON molecules(ssIndex1);
CREATE INDEX IF NOT EXISTS ssIndex2 ON molecules(ssIndex2);
CREATE INDEX IF NOT EXISTS ssIndex3 ON molecules(ssIndex3);
CREATE INDEX IF NOT EXISTS ssIndex4 ON molecules(ssIndex4);
CREATE INDEX IF NOT EXISTS ssIndex5 ON molecules(ssIndex5);
CREATE INDEX IF NOT EXISTS ssIndex6 ON molecules(ssIndex6);
CREATE INDEX IF NOT EXISTS ssIndex7 ON molecules(ssIndex7);
`;

  db.exec(createTable);
  return db;
}
