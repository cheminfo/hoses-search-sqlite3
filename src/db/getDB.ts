import { existsSync, mkdirSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

import { v4 } from '@lukeed/uuid';
import sqLite, { Database } from 'better-sqlite3';
import Postgrator from 'postgrator';

const __dirname = dirname(fileURLToPath(import.meta.url));

let db: Database;

export async function getDB(): Promise<Database> {
  if (!db?.open) {
    const path = new URL('../../sqlite/', import.meta.url).pathname;
    if (!existsSync(path)) {
      mkdirSync(path);
    }
    db = sqLite(join(path, 'db.sqlite'));
    // https://www.sqlite.org/wal.html
    // Activating WAL mode allows to get a speed improvement of 100x !!!
    db.pragma('journal_mode = WAL');
  }
  await prepareDB(db);
  return db;
}

export function getAllTableInfo(db: Database) {
  const sql = `
SELECT name, sql
FROM sqlite_master
WHERE type='table';`;
  const stmt = db.prepare(sql);
  const rows = stmt.all();
  return rows;
}

export async function getTempDB(): Promise<Database> {
  const tempDB = sqLite(`file:${v4()}?mode=memory`);
  await prepareDB(tempDB);
  return tempDB;
}

export async function prepareDB(db: Database) {
  const postgrator = new Postgrator({
    migrationPattern: join(__dirname, 'migrations/*'),
    driver: 'sqlite3',
    execQuery: (query) => {
      return new Promise((resolve, reject) => {
        const stmt = db.prepare(query);
        try {
          const rows = stmt.all();
          resolve({ rows });
        } catch (error) {
          //@ts-expect-error This is a hack to avoid the error
          if (error.message.includes('This statement does not return data')) {
            stmt.run();
            resolve({ rows: [] });
          }
          throw error;
        }
      });
    },
    execSqlScript: (sqlScript) => {
      return new Promise((resolve, reject) => {
        db.exec(sqlScript);
        resolve();
      });
    },
  });
  await postgrator.migrate();
}
