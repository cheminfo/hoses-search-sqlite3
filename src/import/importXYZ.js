import { getEntries } from '../qm9/utils/getEntries.js';
import { readFileSync } from 'fs';

export function importXYZ(xyzFileName) {
  const xyzData = readFileSync(xyzFileName, 'utf-8');
  // console.log(xyzData);
  const entries = getEntries(xyzData);
  // console.log(entries);
  // split XYZ
  return entries;
}
