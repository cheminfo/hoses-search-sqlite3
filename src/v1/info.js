import debugLibrary from 'debug';

const debug = debugLibrary('getInfoFromSmiles');

export default function info(fastify) {
  fastify.route({
    url: '/v1/info',
    method: ['GET', 'POST'],
    handler: getInfo,
    scheme: {
      summary: '$retrieve information from a molfile',
      description: '',
    },
  });
}

async function getInfo(request, response) {
  const body = request.query;
  try {
    const result = {};
    return await response.send({ result });
  } catch (error) {
    debug(`Error: ${error.stack}`);
    return response.send({ result: {}, log: error.toString() });
  }
}

export function getAlgorithmsWithContact(db) {
  const infoStmt = db.prepare(`
  SELECT algorithmID, name, version, description, algorithms.contactID, mail
  FROM algorithms
  INNER JOIN contacts ON algorithms.contactID = contacts.contactID`);
  return infoStmt.all();
}

export function getAtomCountPerAlgorithm(db) {
  const AtomCountstmt = db.prepare(`
SELECT energies.algorithmID, algorithms.name, algorithms.version, COUNT(energies.atomID) AS atomCount
FROM algorithms
INNER JOIN energies ON algorithms.algorithmID = energies.algorithmID
GROUP BY energies.algorithmID`);
  return AtomCountstmt.all();
}

export function getHoseCountPerAlgorithm(db) {
  const hoseCountStmt = db.prepare(`
SELECT energies.algorithmID, COUNT(hoseCodes.hoseID) AS hoseCount
FROM atoms
INNER JOIN hoseCodes ON hoseCodes.atomID = atoms.atomID
INNER JOIN energies ON energies.atomID = atoms.atomID
GROUP BY energies.algorithmID`);
  return hoseCountStmt.all();
}

export function getEntryCountPerAlgorithm(db) {
  const EntryCountStmt = db.prepare(`
SELECT energies.algorithmID, COUNT(entries.entryID) AS entryCount
FROM atoms
INNER JOIN entries ON atoms.entryID = entries.entryID
INNER JOIN energies ON atoms.atomID = energies.atomID
GROUP BY energies.algorithmID`);
  return EntryCountStmt.all();
}
