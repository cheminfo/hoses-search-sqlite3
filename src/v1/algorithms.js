import debugLibrary from 'debug';

import { getDB } from '../db/getDB.js';

const debug = debugLibrary('getInfoFromSmiles');

export default function algorithms(fastify) {
  fastify.route({
    url: '/v1/algorithms',
    method: ['GET', 'POST'],
    handler: getAlgorithms,
    schema: {
      summary: '$retrieve information from a molfile',
      description: '',
    },
  });
}

async function getAlgorithms(request, response) {
  try {
    const db = await getDB();
    const algorithms = getAlgorithmsFromDB(db);
    return response.send(algorithms);
  } catch (error) {
    return response.send({ log: error.toString() });
  }
}

export function getAlgorithmsFromDB(db) {
  try {
    const stmt = db.prepare(`SELECT * from algorithms`);
    return stmt.all();
  } catch (error) {
    throw new Error(error);
  }
}
