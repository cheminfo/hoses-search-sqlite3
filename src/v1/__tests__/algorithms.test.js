import { readFileSync } from 'node:fs';

import { test, expect } from 'vitest';

import { getTempDB } from '../../db/getDB.js';
import { importXYZ } from '../../import/importXYZ.js';
import { getAlgorithmsFromDB } from '../algorithms.js';

import { testDataProperties } from './data/test.info.js';

test('getAlgorithms', async () => {
  const contents = [
    readFileSync(new URL('data/test_1.xyz', import.meta.url)),
    readFileSync(new URL('data/test_2.xyz', import.meta.url)),
    readFileSync(new URL('data/test_3.xyz', import.meta.url)),
  ];
  const db = await getTempDB();
  for (let i = 0; i < contents.length; i++) {
    await importXYZ(contents[i], db, testDataProperties[i]);
  }
  const algorithms = getAlgorithmsFromDB(db);
  expect(algorithms).toMatchSnapshot('algorithms');
});
