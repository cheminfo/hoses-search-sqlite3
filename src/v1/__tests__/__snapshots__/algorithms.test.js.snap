// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getAlgorithms > algorithms 1`] = `
[
  {
    "algorithmID": 1,
    "contactID": 1,
    "data_type": null,
    "description": "GW_charged test 1",
    "name": "GW_charged",
    "version": "1.0.0",
  },
  {
    "algorithmID": 2,
    "contactID": 1,
    "data_type": null,
    "description": "dKS_neutral test 1",
    "name": "dKS_neutral",
    "version": "1.0.0",
  },
  {
    "algorithmID": 3,
    "contactID": 1,
    "data_type": null,
    "description": "dKS_charged test 1",
    "name": "dKS_charged",
    "version": "1.0.0",
  },
  {
    "algorithmID": 4,
    "contactID": 2,
    "data_type": null,
    "description": "GW_charged test 2",
    "name": "GW_charged",
    "version": "1.5.0",
  },
  {
    "algorithmID": 5,
    "contactID": 2,
    "data_type": null,
    "description": "dKS_charged test 2",
    "name": "dKS_charged",
    "version": "1.2.0",
  },
  {
    "algorithmID": 6,
    "contactID": 2,
    "data_type": null,
    "description": "dKS_neutral test 2",
    "name": "dKS_neutral",
    "version": "1.6.0",
  },
  {
    "algorithmID": 7,
    "contactID": 3,
    "data_type": null,
    "description": "GW_charged test 3",
    "name": "GW_charged",
    "version": "2.0.0",
  },
  {
    "algorithmID": 8,
    "contactID": 3,
    "data_type": null,
    "description": "dKS_charged test 3",
    "name": "dKS_charged",
    "version": "3.0.0",
  },
  {
    "algorithmID": 9,
    "contactID": 3,
    "data_type": null,
    "description": "dKS_neutral test 3",
    "name": "dKS_neutral",
    "version": "4.0.0",
  },
]
`;
